Fix set_default_cursor_shape interaction with Control nodes

capnm opened the pull request:
Do not call `set_cursor_shape` when the cursor is inside the `Control` node.

Make it work for x11 in MOUSE_MODE_CONFINED.

Test demo [v2 - added a custom cursor shape test]
[pr22070-set_default_cursor_shape.zip](https://github.com/godotengine/godot/files/2385518/pr22070-set_default_cursor_shape.zip)

![peek4](https://user-images.githubusercontent.com/4047289/45584327-f0a7bb00-b8d1-11e8-8578-90c9929d57ac.gif)

guilhermefelipecgs commented on main/input_default.cpp:
```diff
@@ -598,7 +598,8 @@ Input::CursorShape InputDefault::get_default_cursor_shape() {
 	default_shape = p_shape;
-	OS::get_singleton()->set_cursor_shape((OS::CursorShape)p_shape);
+	// Triggers _gui_input_event, which then does the right thing.
+	warp_mouse_position(mouse_pos);
```
This solution is not too elegant and I'm not sure if this will works for windows and mac, windows's doc don't say if `SetCursorPos` will trigger event and mac docs say that `CGWarpMouseCursorPosition` will not.
https://docs.microsoft.com/en-us/windows/desktop/api/winuser/nf-winuser-setcursorpos
https://developer.apple.com/documentation/coregraphics/1456387-cgwarpmousecursorposition?language=objc

guilhermefelipecgs replied:
Probably I can test on windows but we will need someone to test on mac.

capnm replied:
Any better Idea how to get the event, or how to check we are not over Control node? Alternatively we could document that the shape change happens after a mouse motion, but that's also not nice.

guilhermefelipecgs replied:
Well, to get this working for all platforms, an alternative could be call "send event" method in the last line of `set_cursor_shape` for all platforms to dispatch an mouse motion event.

[windows] https://docs.microsoft.com/en-us/windows/desktop/api/winuser/nf-winuser-sendinput
[x11] https://www.x.org/archive/X11R7.5/doc/man/man3/XSendEvent.3.html
[osx] https://developer.apple.com/documentation/appkit/nswindow/1419228-sendevent?language=objc

capnm replied:
> in the last line of set_cursor_shape

That won't work. The actual problem is that `XDefineCursor` immediately changes the cursor shape
https://www.x.org/releases/X11R7.5/doc/man/man3/XUndefineCursor.3.html

`set_cursor_shape` is already called in` _gui_input_event` exactly as we need it = only outside the Control nodes.

https://github.com/godotengine/godot/blob/af5c30acb2eee54ce78e3dc713eea97bd3ddc18d/scene/main/viewport.cpp#L1901-L1904

All we have to do is trigger the GUI event, otherwise nothing happens if you call `set_default_cursor_shape` with the cursor lying in the viewport - well, until you move the mouse.

guilhermefelipecgs replied:
Use this then:
```
Ref<InputEventMouseMotion> mm;
mm.instance();
mm->set_position(mouse_pos);
parse_input_event(mm);
```
instead of `warp_mouse_position(mouse_pos);`, this should works for all platforms.

capnm replied:
Excellent, tested on Linux, cross-compiled Windows and 3D  (^_^).

![peek6](https://user-images.githubusercontent.com/4047289/45589706-87a06180-b92a-11e8-8941-b403472d4e7a.gif)

guilhermefelipecgs replied:
:+1:

guilhermefelipecgs commented on platform/x11/os_x11.cpp:
```diff
@@ -2524,13 +2524,16 @@ void OS_X11::set_cursor_shape(CursorShape p_shape) {
-		if (cursors[p_shape] != None)
+	}
+
+	if (mouse_mode == MOUSE_MODE_VISIBLE || mouse_mode == MOUSE_MODE_CONFINED) {
```
Changes in this file is right.

capnm committed:
Fix set_default_cursor_shape interaction with Control nodes

Do not call `set_cursor_shape` when the cursor is inside the `Control` node.

Make it work for x11 in MOUSE_MODE_CONFINED.

guilhermefelipecgs commented:
> Do not call `set_cursor_shape` when the cursor is inside the `Control` node.

To simplify for who will merge it, this PR just prevent cursor shape to change immediately when cursor is over some control node because it should only affect viewport.

before:
![peek 15-09-2018 17-32](https://user-images.githubusercontent.com/1387165/45590377-56656880-b90d-11e8-8441-936ab8c07492.gif)
after:
![peek 15-09-2018 17-31](https://user-images.githubusercontent.com/1387165/45590376-56656880-b90d-11e8-8a64-552c88d5d829.gif)
